### DISCUSSION TITLE
After Powershell `Set-Date`, Windows is resetting to live Date shortly

### DISCUSSION THREAD

**Discussion by Gustl22** [actions/virtual-environments#8105] [General questions] [answered]
### Description

In order to install a certificate for a virtual audio device (see #2528, which is also not resolved completely), I need to reset the system date time.

But after setting the time with:
```powershell
Set-Date (Get-Date "2023-07-04 12:00:00");
```
It is reset immediately after that. See [the run](https://github.com/bluefireteam/audioplayers/actions/runs/5854052282/job/15869226966).

I already asked this in the Microsoft QA: https://learn.microsoft.com/en-us/answers/questions/1345225/powershell-set-date-is-resetting-to-its-original-v

They proposed, that the error is in the system host, which constantly trying to keep the date up to date. Is there a way to disable resetting the system time from the host side?

### Platforms affected

- [ ] Azure DevOps
- [X] GitHub Actions - Standard Runners
- [ ] GitHub Actions - Larger Runners

### Runner images affected

- [ ] Ubuntu 20.04
- [ ] Ubuntu 22.04
- [ ] macOS 11
- [ ] macOS 12
- [ ] macOS 13
- [ ] Windows Server 2019
- [X] Windows Server 2022

### Image version and build link

Version: 20230804.1.0

### Is it regression?

I don't think so

### Expected behavior

If setting the time with `Set-time`, then keep the time on the guest system.

### Actual behavior

The time is resetting immediately after setting the custom time, see "Show timestamps" on a regular GH action run.

### Repro steps

```powershell
net stop w32time; W32tm /unregister; Set-Date (Get-Date "2023-07-04 12:00:00")
# Do stuff....
# ...date is reset to current DateTime before finishing the tasks
# Reset the time if wanted:
# W32tm /register; net start w32time; w32tm /resync /force; $currentDate = Get-Date; Write-Host "Current System Date: $currentDate"

**Comment by shamil-mubarakshin** [comment]
Hello @Gustl22, thanks for reporting.
We will take a look

**Comment by ilia-shipitsin** [comment, marked as answer]
windows runs on top of Hyper-V. there's dedicated Hyper-V sync service.

please try the following

```
Set-Service -Name vmictimesync -Status stopped -StartupType disabled
Set-ItemProperty HKLM:\SYSTEM\CurrentControlSet\services\W32Time\Parameters -Name 'Type' -Value 'NoSync'
```

**Reply by Gustl22** [reply]
Thank you very much, I'll try it out ;D

**Reply by Gustl22** [reply]
I tried it and this gives the following error:
```powershell
Set-Service : Cannot stop service 'Hyper-V Time Synchronization Service (vmictimesync)' because it is dependent on 
other services.
At D:\a\_temp\cb72c722-44e9-48b2-b9b0-bdc9172b6d77.ps1:5 char:1
+ Set-Service -Name vmictimesync -Status stopped -StartupType disabled
```

See https://github.com/bluefireteam/audioplayers/actions/runs/5889952997/job/15974114183

**Reply by ilia-shipitsin** [reply]
interesting,

I'm able to stop it

https://github.com/ilia-shipitsin/dependent/actions/runs/5890133706/job/15974638143

**Reply by ilia-shipitsin** [reply]
can you please add `Get-Service -Name vmictimesync -DependentServices` to your pipeline ?

**Reply by Gustl22** [reply]
Run failed with `Get-Service` and got no output:
https://github.com/bluefireteam/audioplayers/actions/runs/5890883319/job/15976877169

Also tried with piping `| Stop-Service -Force`:
https://github.com/bluefireteam/audioplayers/actions/runs/5890942684/job/15977044302

Also changing the to push trigger type didn't succeed:
https://github.com/bluefireteam/audioplayers/actions/runs/5891042556/job/15977327907

**Reply by Gustl22** [reply]
Setting 
```
shell: powershell
```
triggers the error. Do you know why? Didn't know that powershell is the default instead of `cmd`? But setting the shell shouldn't make a difference, should it?

**Reply by ilia-shipitsin** [reply]
I wonder what's going on with `shell: powershell`, I'm able to repro your failure, but I do not see anything on ` Get-Service -Name vmictimesync -DependentServices`

Specifying `shell: pwsh` works for me. You can move "time sync disable" to separate job and specify `shell: pwsh` only for that job

**Reply by Gustl22** [reply]
I removed setting the shell completely, which succeeded. But in this case there shouldn't be a difference in setting `pwsh` or `powershell`. This may should be investigated in a ticket...

Setting the time works now I think :) Thank you for sacrificing your time ;D

**Reply by ilia-shipitsin** [reply]
working approach (GHA)

```
      - name: disable time sync
        run: |
          Set-Service -Name vmictimesync -Status stopped -StartupType disabled
          Set-ItemProperty HKLM:\SYSTEM\CurrentControlSet\services\W32Time\Parameters -Name 'Type' -Value 'NoSync'
        shell: pwsh
```

**Comment by Gustl22** [comment]
For other readers: This is the command to reenable the Hyper-V time sync:

```powershell
Set-ItemProperty HKLM:\SYSTEM\CurrentControlSet\services\W32Time\Parameters -Name 'Type' -Value 'NTP'
Set-Service -Name vmictimesync -Status running -StartupType automatic
```
